[DOCUMENTATION][feature/otx] Explanation section. PR 2

kprokofi opened the pull request:
- [x] Fix classification sections (move metrics to model sections, add some refs)
- [x] Semantic segmentation (+ Semi-SL)
- [x] Instance segmentation
- [x] Auto-config (simple one)

kprokofi committed:
added semantic segmentation. Instance segmentation. Auto-config.

kprokofi committed:
fix conflicts

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,36 @@
 
-Coming soon...
\ No newline at end of file
+Auto configuration for a deep learning framework refers to the process of automatically determining the most appropriate settings for the framework's parameters, such as specific algorithm and set of hyperparameters, based on the dataset and the specific task at hand.
```
```suggestion
Auto-configuration for a deep learning framework means the automatic finding of the most appropriate settings for the training parameters, based on the dataset and the specific task at hand.
```

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,36 @@
-Coming soon...
\ No newline at end of file
+Auto configuration for a deep learning framework refers to the process of automatically determining the most appropriate settings for the framework's parameters, such as specific algorithm and set of hyperparameters, based on the dataset and the specific task at hand.
+Auto configuration can help to save time, ease the process of interaction with OTX CLI and give a better baseline for the given dataset.
```
```suggestion
Auto-configuration can help to save time, ease the process of interaction with OTX CLI and give a better baseline for the given dataset.
```

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,36 @@
+Auto configuration can help to save time, ease the process of interaction with OTX CLI and give a better baseline for the given dataset.
+
+At this end, we developed a simple auto-configuration functionality to ease the process of training and validation utilizing our framework.
+Basically, to start training and obtain a good baseline with best trade off between accuracy and speed we need to pass only a dataset in the right format
```
```suggestion
Basically, to start the training and obtain a good baseline with the best trade-off between accuracy and speed we need to pass only a dataset in the right format
```

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,36 @@
+
+    $ otx train --train-data-root <path_to_data_root>
+
+Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated to OTX will handle it on its own.
```
```suggestion
Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated into OTX will handle it on its own.
```

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,36 @@
+
+Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated to OTX will handle it on its own.
+It will recognize the task by analizing the dataset and if there is no splits for the validation - Datumaro will do random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.
+Currently, Datumaro auto-split feature supports 3 tasks: multi-class classification, detection and semantic segmentation.
```
Can you highlight task names using bold or italic font, please?

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,36 @@
+
+Supported dataset formats for each task:
+
+    - classificaiton: Imagenet, COCO (multi-label), :ref:`custom hierarchical <hierarchical_dataset>`
+    - object detection: COCO, Pascal-VOC, YOLO
+    - semantic segmentation: common semantic segmentation, Pascal-VOC, cityscapes, ADE20k
+    - action classification: CVAT
+    - action : CVAT
+    - anomaly classificaiton: MVTec
+    - anomaly detection: MVTec
+    - anomaly segmentation: MVTec
+    - instance segmentation: COCO, Pascal-VOC
```
I'd suggest removing indents here, because it's rendered similarly to Note field
![image](https://user-images.githubusercontent.com/33455576/214813396-f89031c8-cf09-4ce2-bec1-1ce264fc7ddb.png)

GalyaZalesskaya replied:
I would also add links to these formats, especially because you already have them in Segmentation section `COCO <https://cocodataset.org/#format-data>` and `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,36 @@
+    - anomaly segmentation: MVTec
+    - instance segmentation: COCO, Pascal-VOC
+
+If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with additional CLI option:
```
```suggestion
If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option:
```

GalyaZalesskaya replied:
Also, I'd personally put this into Note field

GalyaZalesskaya commented on docs/source/explanation/additional_features/models_optimization.rst:
```diff
@@ -14,4 +14,4 @@ Models Optimization
 | The main recommendation is to start with post-training compression and use NNCF compression during training if you are not satisfied with the results.
-| Please, refer to our `dedicated tutorials <LINK>` on how to optimize your model using POT or NNCF.
\ No newline at end of file
+| Please, refer to our `dedicated tutorials <../../tutorials/base/how_to_train>` on how to optimize your model using POT or NNCF.
```
```suggestion
| Please, refer to our :doc:`dedicated tutorials <../../tutorials/base/how_to_train>` on how to optimize your model using POT or NNCF.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/classification/multi_class_classification.rst:
```diff
@@ -78,14 +78,25 @@ We support the following ready-to-use model templates:
 
 Besides this, we support public backbones from `torchvision <https://pytorch.org/vision/stable/index.html>`_, `pytorchcv <https://github.com/osmr/imgclsmob>`_, `mmcls <https://github.com/open-mmlab/mmclassification>`_ and `OpenVino Model Zoo <https://github.com/openvinotoolkit/open_model_zoo>`_.
-Please, refer to the `tutorial <N/A>`_ how to customize models and run public backbones.
+Please, refer to the :doc:`tutorial <../../tutorials/advanced/backbones.rst>`_ how to customize models and run public backbones.
```
```suggestion
Please, refer to the :doc:`tutorial <../../tutorials/advanced/backbones>` how to customize models and run public backbones.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
 
+Semantic segmentation is a computer vision task in which an algorithm assigns a label or class to each pixel in an image.
+For example, semantic segmentation can be used to identify the boundaries of different objects in an image, such as cars, buildings, and trees.
+The output of semantic segmentation is typically an image where each pixel is colored with a different color or label depending on its class.
```
New paragraph (double enter). Also, it'd be nice to add an image, visualizing semantic segmentation inference, what do you think?

kprokofi replied:
good idea!

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
+
+.. _semantic_segmentation_supervised_pipeline:
+
+- ``Augmentations``: Besides basic augmentations like random flip, randomly rotate and random crop, we use mixing images technique with different `photometric distortions <https://mmsegmentation.readthedocs.io/en/latest/api.html#mmseg.datasets.pipelines.PhotoMetricDistortion>`_.
```
```suggestion
- ``Augmentations``: Besides basic augmentations like random flip, random rotate, and random crop, we use mixing images technique with different `photometric distortions <https://mmsegmentation.readthedocs.io/en/latest/api.html#mmseg.datasets.pipelines.PhotoMetricDistortion>`_.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
+
+- ``Loss function``: We use `max-pooling loss <https://arxiv.org/pdf/1704.02966.pdf>`_ with standard `Cross Entropy Loss <https://en.wikipedia.org/wiki/Cross_entropy>`_ as the base for handling imbalanced training data distributions. This technique helps re-weighting the contributions of each pixel based on their loss value, targeting under-performing classification categories.
+
+- Additionally, we use **early stopping** to add adaptability to the training pipeline and prevent overfitting.
```
```suggestion
- Additionally, we use the **early stopping** to add adaptability to the training pipeline and prevent overfitting.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
+
+    $ otx {train, optimize} <model_template> --train-data-root <path_to_data_root>
+
+If we have a dataset in the format below:
```
```suggestion
To do that, we need to have a dataset with the following format:
```

kprokofi replied:
That is not what I meant. I rephrased it a bit

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
+                ...
+                └── N.png
+
+Where we have images and segmentation annotations masks with the same naming. Then we also can start training directly passing the training and validation root paths as well as the training and validation segmentation masks paths like in the following command line:
```
```suggestion
There we have images and segmentation annotations masks with the same naming. 

We also can start training directly by passing the training and validation root paths as well as the training and validation segmentation masks path:

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
+
+.. note::
+
+Please, refer to our :doc:`dedicated tutorial <../../../tutorials/base/how_to_train/semantic_segmentation>` for more information on how to train, validate and optimize the semantic segmentation model.
```
You need to add an indent here for a proper rendering

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
+| `Custom_Semantic_Segmentation_Lite-HRNet-x-mod3_OCR <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_x_mod3/template.yaml>`_                      | Lite-HRNet-x-mod3      | 13.97               | 6.4             |
++------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------+---------------------+-----------------+
+
+All of these models are members of the same `Lite-HRNet <https://arxiv.org/abs/2104.06403>`_ backbones family. They differ in the trade-off between accuracy and inference/training speed. Lite-HRNet-x-mod3 is the template with heavy-size architecture for accurate predictions but long training.
```
```suggestion
All of these models are members of the same `Lite-HRNet <https://arxiv.org/abs/2104.06403>`_ backbones family. They differ in the trade-off between accuracy and inference/training speed. Lite-HRNet-x-mod3 is the template with heavy-size architecture for accurate predictions but it requires long training.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
 
-To be added soon
+Semi-supervised semantic segmentation is a variation of the semantic segmentation task, where only a subset of the data is labeled.
+The goal is to train a model to predict the class labels of the pixels in an image, using both labeled and unlabeled data.
```
```suggestion
The goal is to train a model to predict the class labels of the pixels in an image, using both **labeled and unlabeled data**.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
+In this way, the model can make use of both labeled data to learn the class labels, and unlabeled data to learn the underlying structure of the image.
+This can be useful in situations where acquiring labeled data is expensive or time-consuming.
+
+To solve this problem we use the Mean Teacher algorithm. The basic idea of this approach is to use two models during training: a "student" model, which is the main model being trained, and a "teacher" model, which acts as a guide for the student model.
```
Maybe a link for a Mean Teacher?

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,142 @@
+
+We utilize the same core algorithm's parameters as for the :ref:`supervised pipeline <semantic_segmentation_supervised_pipeline>`. The main difference is to use of different augmentation pipelines for the labeled and unlabeled data.
+We use only basic augmentations (random flip, random rotate, random crop) for the labeled data and more severe for the unlabeled (the same as we use for the :ref:`supervised pipeline <semantic_segmentation_supervised_pipeline>`).
+It helps for better generalization and prevents unnecessary overfitting on the pseudo-labels generated by the teacher model.
```
```suggestion
It helps with a better generalization and prevents unnecessary overfitting on the pseudo-labels generated by the teacher model.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,69 @@
 =====================
 
+| Instance segmentation is a computer vision task that involves identifying and segmenting individual objects within an image.
+| It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows for more detailed information about the object, such as its shape and location, to be extracted.
```
```suggestion
| It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows for more detailed information about the object, such as its shape and location, to be extracted.
```
```suggestion
| It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows getting more detailed information about the object, such as its shape and location, to be extracted.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,69 @@
 
+| Instance segmentation is a computer vision task that involves identifying and segmenting individual objects within an image.
+| It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows for more detailed information about the object, such as its shape and location, to be extracted.
+| Instance segmentation is commonly used in applications such as self-driving cars, robotics, and image-editing software.
```
Double enter after this line to start a new paragraph. I'd also suggest using image for visualization like [this](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/instance-segmentation-security-0002/assets/instance-segmentation-security-0002.png)

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,69 @@
+
+- ``Loss functions``: For the bounding box regression we use **L1 Loss** (the sum of the absolute differences between the ground truth value and the predicted value), `Cross Entropy Loss <https://en.wikipedia.org/wiki/Cross_entropy>`_ for the categories classification and segmentation masks prediction.
+
+- Additionally, we use  **Exponential Moving Average (EMA)** for the model's weights and **early stopping** to add adaptability to the training pipeline and prevent overfitting.
```
```suggestion
- Additionally, we use  the **Exponential Moving Average (EMA)** for the model's weights and the **early stopping** to add adaptability to the training pipeline and prevent overfitting.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,69 @@
+| `Custom_Counting_Instance_Segmentation_MaskRCNN_ResNet50 <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/detection/configs/instance_segmentation/resnet50_maskrcnn/template.yaml>`_                | MaskRCNN-ResNet50          | 533.80              | 177.90          |
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------------------+-----------------+
+
+``MaskRCNN-ResNet50`` uses `ResNet-50 <https://arxiv.org/abs/1512.03385>`_ as the backbone network for the image features extraction. It has more parameters and Flops and needs more time to train, meanwhile providing superior performance in terms of accuracy. ``MaskRCNN-EfficientNetB2B`` uses `EfficientNet-B2 <https://arxiv.org/abs/1905.11946>`_ as the backbone network. It is a good trade off bee ...
```
```suggestion
``MaskRCNN-ResNet50`` uses `ResNet-50 <https://arxiv.org/abs/1512.03385>`_ as the backbone network for the image features extraction. It has more parameters and FLOPs and needs more time to train, meanwhile providing superior performance in terms of accuracy. ``MaskRCNN-EfficientNetB2B`` uses `EfficientNet-B2 <https://arxiv.org/abs/1905.11946>`_ as the backbone network. It is a good trade-off between accuracy and speed. It is a better choice when training time and computational cost are in priority.
```

GalyaZalesskaya reviewed (commented):
Thank you for such a great explanation. I also have one comment about Semi-SL explanation in Semantic Segmentation. Since it's a common explanation to all tasks, don't you think it's better to move to a standalone page in `Algorithms`? I will not repeat yourself in each task and just make a link to Semi-SL explanation and keep just numbers.

kprokofi commented:
> Thank you for such a great explanation. I also have one comment about Semi-SL explanation in Semantic Segmentation. Since it's a common explanation to all tasks, don't you think it's better to move to a standalone page in `Algorithms`? I will not repeat yourself in each task and just make a link to Semi-SL explanation and keep just numbers.

The problem is that algorithms may vary from task to task and for segmentation it will be MeanTeacher, for classification it will BarlowTwins (for example).
I agree that description of what is the Semi-SL task is common though. Maybe I can think about adding the task description on the main explanation page

GalyaZalesskaya commented:
> The problem is that algorithms may vary from task to task and for segmentation it will be MeanTeacher, for classification it will BarlowTwins (for example). I agree that description of what is the Semi-SL task is common though. Maybe I can think about adding the task description on the main explanation page

What do you think about the idea of making a standalone page with `Semi-SL` algorithms under the `Algorithms` section with Mean teacher, Barrow twins algorithms and make a link to it from Semi-SL section in each task? Because personally I'd avoid repetition of Mean-teacher explanation in `Instance Segmentation` and `Sematic Segmentation`.

Or, as the other option, these explanations could be transferred to common Semi-SL page under the `Segmentation` section

kprokofi commented:
> > The problem is that algorithms may vary from task to task and for segmentation it will be MeanTeacher, for classification it will BarlowTwins (for example). I agree that description of what is the Semi-SL task is common though. Maybe I can think about adding the task description on the main explanation page
> 
> What do you think about the idea of making a standalone page with `Semi-SL` algorithms under the `Algorithms` section with Mean teacher, Barrow twins algorithms and make a link to it from Semi-SL section in each task? Because personally I'd avoid repetition of Mean-teacher explanation in `Instance Segmentation` and `Sematic Segmentation`.
> 
> Or, as the other option, these explanations could be transferred to common Semi-SL page under the `Segmentation` section

I moved description of the training types and algorithms supported by OTX to the index. I removed semi-sl description from task specific page
![image](https://user-images.githubusercontent.com/68103869/214944856-10af81a1-ef95-4493-9c3d-e3adabb261a2.png)

kprokofi committed:
fix auto-config, model_opt

kprokofi committed:
added links to data

kprokofi committed:
added images

kprokofi committed:
reflect comments

kprokofi committed:
move training types description to index

kprokofi committed:
fix grammar

kprokofi commented:
@sungmanc , Could you take a look at the simple-auto config section? Did I write it right? Did I miss smth?
@eugene123tw, Could you take a look at instance segmentation part? Maybe I wrote smth wrong or missed important things. Also, can I ask you to write section about tiling after this PR merged? I'm not an expert in that. We need to expose users what it is for, when it is useful, some short insights about internal kitchen
@JihwanEom Could you review semantic segmentation as well? :)

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+
+- ``Optimizer``: We use `Adam <https://arxiv.org/abs/1412.6980>`_ optimizer with weight decay set to zero and gradient clipping with maximum quadratic norm equals to 40.
+
+- ``Learning rate schedule``: For scheduling training process we use **ReduceLrOnPlataeu** with linear learning rate warmup for 80 iterations. This method monitors a target metric (in our case we use metric on the validation set) and if no improvement is seen for a 'patience' number of epochs, the learning rate is reduced.
```
```suggestion
- ``Learning rate schedule``: For scheduling training process we use **ReduceLROnPlateau** with linear learning rate warmup for 100 iterations. This method monitors a target metric (in our case we use metric on the validation set) and if no improvement is seen for a 'patience' number of epochs, the learning rate is reduced.
```

https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/template.yaml#L36

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+
+- ``Learning rate schedule``: For scheduling training process we use **ReduceLrOnPlataeu** with linear learning rate warmup for 80 iterations. This method monitors a target metric (in our case we use metric on the validation set) and if no improvement is seen for a 'patience' number of epochs, the learning rate is reduced.
+
+- ``Loss function``: We use `max-pooling loss <https://arxiv.org/pdf/1704.02966.pdf>`_ with standard `Cross Entropy Loss <https://en.wikipedia.org/wiki/Cross_entropy>`_ as the base for handling imbalanced training data distributions. This technique helps re-weighting the contributions of each pixel based on their loss value, targeting under-performing classification categories.
```
We are not using `max-pooling loss`: https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/model.py#L43

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
 
-To be added soon
+For the dataset handling inside OTX we use `Dataset Management Framework (Datumaro) <https://github.com/openvinotoolkit/datumaro>`_. At this end we support `ADE20K <https://openvinotoolkit.github.io/datumaro/docs/formats/ade20k2020/>`_, `Cityscapes <https://openvinotoolkit.github.io/datumaro/docs/formats/cityscapes/>`_, `Pascal VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_v ...
+Running training with these dataset formats is very simple. We just need to pass a path to the root folder and desired model template to start training:
```
How about
```suggestion
If you organized supported dataset format, starting training will be very simple. We just need to pass a path to the root folder and desired model template to start training:
```

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+
+    $ otx {train, optimize} <model_template> --train-data-root <path_to_data_root>
+
+When we have a dataset in the format below, where we have images and segmentation annotations masks with the same naming:
```
How about
```suggestion
Image-annotation pairs are supposed to have same name as below:
```

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+    data
+        ├── images
+            ├── train
+                ├── 0.png
```
Is `png` supported extender for images? I've used only `jpg` for images, `png` for annotations.

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+                ...
+                └── N.png
+
+We also can start training directly passing the training and validation root paths as well as the training and validation segmentation masks paths like in the following command line:
```
```suggestion
We can also start training by passing the training and validation data paths manually:
```

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
++==============================================================================================================================================================================================================================+========================+=====================+=================+
+| `Custom_Semantic_Segmentation_Lite-HRNet-s-mod2_OCR <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_s_mod2/template.yaml>`_                      | Lite-HRNet-s-mod2      | 1.82                | 3.5             |
++------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------+---------------------+-----------------+
+| `Custom_Semantic_Segmentation_Lite-HRNet-18_OCR <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/template.yaml>`_                              | Lite-HRNet-18          | 3.45                | 4.5             |
```
Hmm, Lite-HRNet family in OTX does not have OCR head now, so the complexity and model size maybe inaccurate.

kprokofi replied:
That's a good point if no one recomputed the model size - we need to do it once again. I will check it.

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
++------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------+---------------------+-----------------+
+| `Custom_Semantic_Segmentation_Lite-HRNet-18_OCR <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/template.yaml>`_                              | Lite-HRNet-18          | 3.45                | 4.5             |
++------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------+---------------------+-----------------+
+| `Custom_Semantic_Segmentation_Lite-HRNet-18-mod2_OCR <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/template.yaml>`_                    | Lite-HRNet-18-mod2     | 3.63                | 4.8             |
```
There is no reason to have different complexity/model size compared with `Lite-HRNet-18`, `18-mod2` has exactly same architecture, only training scheme is different. (No freeze layers, uses adam instead of SGD).

kprokofi replied:
But we have two different templates. I see the solution simply not mention the first version of the Lite-HRNet-18. I think it will only misslead a user

JihwanEom replied:
Oh yes, I agree! My concern will be resolved absolutely after updating complexity and model size for each templates. Let's keep one row for one template.

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
++------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------+---------------------+-----------------+
+
+All of these models are members of the same `Lite-HRNet <https://arxiv.org/abs/2104.06403>`_ backbones family. They differ in the trade-off between accuracy and inference/training speed. Lite-HRNet-x-mod3 is the template with heavy-size architecture for accurate predictions but it requires long training.
+Whereas the Lite-HRNet-s-mod2 is the lightweight architecture for fast inference and training. It is the best choice for the scenario of a limited amount of data. The Lite-HRNet-18 model is the middle-sized architecture for the balance between fast inference and training time.
```
```suggestion
Whereas the Lite-HRNet-s-mod2 is the lightweight architecture for fast inference and training. It is the best choice for the scenario of a limited amount of data. The Lite-HRNet-18-mod2 model is the middle-sized architecture for the balance between fast inference and training time.
```

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+All of these models are members of the same `Lite-HRNet <https://arxiv.org/abs/2104.06403>`_ backbones family. They differ in the trade-off between accuracy and inference/training speed. Lite-HRNet-x-mod3 is the template with heavy-size architecture for accurate predictions but it requires long training.
+Whereas the Lite-HRNet-s-mod2 is the lightweight architecture for fast inference and training. It is the best choice for the scenario of a limited amount of data. The Lite-HRNet-18 model is the middle-sized architecture for the balance between fast inference and training time.
+
+In the table below the `Dice score <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ on some academic datasets using our :ref:`supervised pipeline <semantic_segmentation_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 512x512 image crop resolution, for other hyperparameters, please, refer to the related template. We ...
```
```suggestion
In the table below the `Dice score <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ on some academic datasets using our :ref:`supervised pipeline <semantic_segmentation_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 512x512 image crop resolution, for other hyperparameters, please, refer to the related template. We trained each model with single Nvidia GeForce RTX3090.
```

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+In the table below the `Dice score <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ on some academic datasets using our :ref:`supervised pipeline <semantic_segmentation_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 512x512 image crop resolution, for other hyperparameters, please, refer to the related template. We ...
+
++-----------------------+--------------+------------+-----------------+
+| Model name            | ADE20k       | Cityscapes | Pascal-VOC 2007 |
```
```suggestion
| Model name            | ADE20k       | Cityscapes | Pascal-VOC 2012 |
```

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+It helps with a better generalization and prevents unnecessary overfitting on the pseudo-labels generated by the teacher model.
+
+In the table below the `Dice score <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ with our middle template on some datasets is presented. For comparison, we present the supervised baseline trained on the labeled data only.
+The results were obtained on our templates without any changes. We use 512x512 image resolution, for other hyperparameters, please, refer to the `related template <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/template.yaml>`_. We trained all models on 1 GPU Nvidia GeForce GTX3090.
```
```suggestion
The results were obtained on our templates without any changes. We use 512x512 image resolution, for other hyperparameters, please, refer to the `related template <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/template.yaml>`_. We trained each model with single Nvidia GeForce RTX3090.
```

JihwanEom commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+
+In the table below the `Dice score <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ with our middle template on some datasets is presented. For comparison, we present the supervised baseline trained on the labeled data only.
+The results were obtained on our templates without any changes. We use 512x512 image resolution, for other hyperparameters, please, refer to the `related template <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/template.yaml>`_. We trained all models on 1 GPU Nvidia GeForce GTX3090.
+For `Cityscapes <https://www.cityscapes-dataset.com/>`_ and `Pascal-VOC <http://host.robots.ox.ac.uk/pascal/VOC/voc2007/index.html>`_ we use splits with different ratios of labeled to unlabeled data like `here <https://github.com/charlesCXK/TorchSemiSeg>`_.
```
```suggestion
For `Cityscapes <https://www.cityscapes-dataset.com/>`_ and `Pascal-VOC <http://host.robots.ox.ac.uk/pascal/VOC/voc2012/index.html>`_ we use splits with different ratios of labeled to unlabeled data like `here <https://github.com/charlesCXK/TorchSemiSeg>`_.
```

kprokofi replied:
Is it more common to validate on the 2012 version?

JihwanEom commented on docs/source/explanation/algorithms/classification/multi_class_classification.rst:
```diff
@@ -78,14 +78,25 @@ We support the following ready-to-use model templates:
 
         $ otx find --backbone {torchvision, pytorchcv, mmcls, omz.mmcls}
 
+In the table below the top-1 accuracy on some academic datasets using our :ref:`supervised pipeline <mcl_cls_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 224x224 image resolution, for other hyperparameters, please, refer to the related template. We trained all models on 1 GPU Nvidia GeForce GTX3090.
```
```suggestion
In the table below the top-1 accuracy on some academic datasets using our :ref:`supervised pipeline <mcl_cls_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 224x224 image resolution, for other hyperparameters, please, refer to the related template. We trained each model with single Nvidia GeForce RTX3090.
```

JihwanEom commented on docs/source/explanation/algorithms/classification/multi_label_classification.rst:
```diff
@@ -74,11 +64,27 @@ To use OTX with this format is required to pass annotations files paths and imag
 ******
 We use the same models as for Multi-class classification. Please, refer: :ref:`Classificaiton Models <classificaiton_models>`.
 
+In the table below the `mAP <https://en.wikipedia.org/w/index.php?title=Information_retrieval&oldid=793358396#Average_precision>`_ metrics on some academic datasets using our :ref:`supervised pipeline <ml_cls_supervised_pipeline>` are presented. The results were obtained on our templates without any changes. We use 448x448 image resolution to make the results comparable with academic papers, for  ...
```
```suggestion
In the table below the `mAP <https://en.wikipedia.org/w/index.php?title=Information_retrieval&oldid=793358396#Average_precision>`_ metrics on some academic datasets using our :ref:`supervised pipeline <ml_cls_supervised_pipeline>` are presented. The results were obtained on our templates without any changes. We use 448x448 image resolution to make the results comparable with academic papers, for other hyperparameters, please, refer to the related template. We trained each model with single Nvidia GeForce RTX3090.
```

JihwanEom commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
+
+``MaskRCNN-ResNet50`` uses `ResNet-50 <https://arxiv.org/abs/1512.03385>`_ as the backbone network for the image features extraction. It has more parameters and FLOPs and needs more time to train, meanwhile providing superior performance in terms of accuracy. ``MaskRCNN-EfficientNetB2B`` uses `EfficientNet-B2 <https://arxiv.org/abs/1905.11946>`_ as the backbone network. It is a good trade-off bet ...
+
+In the table below the `mAP <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ metric on some academic datasets using our :ref:`supervised pipeline <instance_segmentation_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 1024x1024 image resolution, for other hyperparameters, please, refer to the related template. We tr ...
```
```suggestion
In the table below the `mAP <https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient>`_ metric on some academic datasets using our :ref:`supervised pipeline <instance_segmentation_supervised_pipeline>` is presented. The results were obtained on our templates without any changes. We use 1024x1024 image resolution, for other hyperparameters, please, refer to the related template. We trained each model with single Nvidia GeForce RTX3090.
```

kprokofi replied:
agree, this sounds  better

JihwanEom reviewed (commented):
Thank you for hard work! I left some comments.

sungmanc commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,37 @@
+Basically, to start the training and obtain a good baseline with the best trade-off between accuracy and speed we need to pass only a dataset in the right format without specifiing anything else:
+
+.. code-block::
+
```
Auto-configuration also supports `otx build --train-data-roots <path_to_data_root>`

kprokofi replied:
Is there a reason to run `otx build` instead of `otx train` for users? If we call `otx build` it means `otx` will create a folder with all components, but the template (model) will be the middle one. So, using auto-config calling otx build is unnecessary, IMO. But, I will add it as a note mentioning that `otx build` can be called to create a folder.

sungmanc commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,37 @@
+Currently, Datumaro auto-split feature supports 3 tasks: **multi-class classification**, **detection** and **semantic segmentation**.
+After dataset preparation the training will be started with the middle-sized template to achieve a competitive accuracy preserving fast inference.
+
+Supported dataset formats for each task:
```
Since it is a description for auto-configuration, how about removing the datasets that can't be used for auto-configurations?

For this, I need to investigate time to analyze which dataset can be used for auto-configuration. Currently, Only 3 datasets (`Imagenet (cls)` `COCO (det)` `Cityscape (seg)`) can be auto-splitted by Datumaro

kprokofi replied:
If we have splits for other datasets and task can we use a simple auto-config?

kprokofi replied:
I mean running `otx train --train-data-roots <path_to_data_root>` with other formats if we have validation folder exactly with the same name as it should be in original dataset format

sungmanc replied:
If we don't use auto-split function, then we can use datasets below:
Classification: ImageNet
Detection: COCO, VOC, YOLO
Segmentation: Common Semantic Segmentation, VOC, Cityscape, ADE20K 2017, 2020

`otx train --train-data-roots <coco_path>`

kprokofi replied:
I moved supported dataset formats into note section. It is a good idea to give an overview about all supported datasets. We also have a note about supported formats by auto-split

sungmanc commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,37 @@
+
+.. note::
+
+    If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option:
```
Want to add some contents

[Before]
If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option:

[After]
If we have a dataset format that occluded with other tasks, for example ``COCO`` format, we should directly emphasize the task type with an additional CLI option. If not, OTX automatically chooses the task type that you might doesn't intend:

sungmanc reviewed (dismissed):
Thanks for the wonderful documentation, I added comment for auto-configuration part.

kprokofi committed:
reflect comments. Added images

kprokofi committed:
Merge branch 'kp/explanation_phase_2' of https://github.com/openvinotoolkit/training_extensions into kp/explanation_phase_2

JihwanEom commented:
@supersoob Could you double check for sem. seg. parts?

kprokofi committed:
remove HRNet-18 mode 1

kprokofi committed:
change perfomance seg values

kprokofi committed:
move datasets formats to note. Fix a link in opt section.

kprokofi commented:
@JihwanEom @GalyaZalesskaya @sungmanc I reflected all the comments

JihwanEom commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,49 @@
+        $ otx build --train-data-root <path_to_data_root>
+
+Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated into OTX will handle it on its own.
+It will recognize the task by analizing the dataset and if there is no splits for the validation - Datumaro will do random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.
```
```suggestion
It will recognize the task by analyzing the dataset and if there is no splits for the validation - Datumaro will do random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/classification/multi_class_classification.rst:
```diff
@@ -15,18 +17,6 @@ For the supervised training we use the following algorithms components:
 - Additionally, we use `No Bias Decay (NBD) <https://arxiv.org/abs/1812.01187>`_ technique and **early stopping** to add adaptability to the training pipeline and prevent overfitting.
 To further enhance the performance of the algorithm in case when we have a small number of data we use `Supervised Contrastive Learning <https://arxiv.org/abs/2004.11362>`_. More specifically, we train a model with two heads: classification head with Influence-Balanced Loss and SupCon head with `Barlow Twins loss <https://arxiv.org/abs/2103.03230>`_.
```
Please, add enter before this paragraph to avoid compilation warning

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+| Supervised Lite-HRNet-18-mod2 (1/16)| N/A          | N/A        | N/A             |
++-------------------------------------+--------------+------------+-----------------+
+| Semi-SL Lite-HRNet-18-mod2 (1/16)        | N/A          | N/A        | N/A             |
++-------------------------------------+--------------+------------+-----------------+
```
Can you please fix the indents in this table? it provokes an error warning during the compilation, so the table isn't shown

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
 Semantic Segmentation
 =====================
 
```
I'm not sure why, but this line is mentioned in warning during compilation (make HTML): ` WARNING: Duplicate explicit target name: "cityscapes".` Can you check it please?

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,49 @@
-Coming soon...
\ No newline at end of file
+Auto-configuration for a deep learning framework means the automatic finding of the most appropriate settings for the training parameters, based on the dataset and the specific task at hand.
+Auto-configuration can help to save time, ease the process of interaction with OTX CLI and give a better baseline for the given dataset.
```
```suggestion
Auto-configuration can help to save time, it eases the process of interaction with OTX CLI and gives a better baseline for the given dataset.
```

GalyaZalesskaya commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,49 @@
+
+.. note::
+
+    Currently, Datumaro auto-split feature supports 3 formats: `Imagenet <https://www.image-net.org/>`_  (**multi-class classification**), `COCO <https://cocodataset.org/#format-data>`_ (**detection**) and `Cityscapes <https://openvinotoolkit.github.io/datumaro/docs/formats/cityscapes/>`_ (**semantic segmentation**).
```
I'd remove highlighting tasks (only here) because links and bold font together look overwhelming.

GalyaZalesskaya commented on docs/source/explanation/algorithms/index.rst:
```diff
@@ -3,6 +3,16 @@ Algorithms
+
+To this end we support:
+
+- **Supervised training**. Supervised learning for computer vision involves training a model on a labeled dataset of images. The model learns to associate specific features in the images with the corresponding labels. This is the most common approach for computer vision tasks such as object detection and image classification.
```
```suggestion
- **Supervised training**. This is the most common approach for computer vision tasks such as object detection and image classification. Supervised learning involves training a model on a labeled dataset of images. The model learns to associate specific features in the images with the corresponding labels. 
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/index.rst:
```diff
@@ -3,6 +3,16 @@ Algorithms
+To this end we support:
+
+- **Supervised training**. Supervised learning for computer vision involves training a model on a labeled dataset of images. The model learns to associate specific features in the images with the corresponding labels. This is the most common approach for computer vision tasks such as object detection and image classification.
+- **Incremental learning**. OTX provides functionality to let the model to learn from new data as it becomes available and improve its performance over time, rather than having to retrain the entire model every time new data is added. OTX supports also the class incremental approach. In this approach, the model is first trained on a set of classes, and then incrementally updated with new classes  ...
```
```suggestion
- **Incremental learning**. This learning approach lets the model train on new data as it becomes available, rather than retraining the entire model on the whole dataset every time new data is added. OTX supports also the class incremental approach. In this approach, the model is first trained on a set of classes, and then incrementally updated with new classes of data, while keeping the previously learned classes' knowledge. The class incremental approach is particularly useful in situations where the number of classes is not fixed and new classes may be added over time.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
 =====================
 
+| Instance segmentation is a computer vision task that involves identifying and segmenting individual objects within an image.
+| It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows getting more detailed information about the object, such as its shape and location, to be extracted.
```
```suggestion
It is a more advanced version of object detection, as it doesn't only detect the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows getting more detailed information about the object, such as its shape and location, to be extracted.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
 
+| Instance segmentation is a computer vision task that involves identifying and segmenting individual objects within an image.
+| It is a more advanced version of object detection, as it not only detects the presence of an object in an image but also segments the object by creating a mask that separates it from the background. This allows getting more detailed information about the object, such as its shape and location, to be extracted.
+| Instance segmentation is commonly used in applications such as self-driving cars, robotics, and image-editing software.
```
```suggestion
Instance segmentation is commonly used in applications such as self-driving cars, robotics, and image-editing software.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
+
+| We solve this problem in the `MaskRCNN <https://arxiv.org/abs/1703.06870>`_ manner. The main idea of Mask R-CNN is to add a branch for predicting an object mask in parallel with the existing branch for bounding box regression and object classification.
+| This is done by using a fully convolutional network (FCN) on top of the feature map generated by the last convolutional layer of the backbone network. The model first generates region proposals, and then uses a RoIAlign layer to align the region proposals with the feature map, then the FCN predicts the class and box offset for each proposal and the mask for each class.
+| The mask branch is trained to predict a binary mask for each object instance, where the mask is aligned with the object's bounding box and has the same size as the region of interest (RoI). The predicted mask is then used to segment the object from the background.
```
```suggestion
We solve this problem in the `MaskRCNN <https://arxiv.org/abs/1703.06870>`_ manner. The main idea of Mask R-CNN is to add a branch for predicting an object mask in parallel with the existing branch for bounding box regression and object classification.
This is done by using a fully convolutional network (FCN) on top of the feature map generated by the last convolutional layer of the backbone network. The model first generates region proposals, and then uses a RoIAlign layer to align the region proposals with the feature map, then the FCN predicts the class and box offset for each proposal and the mask for each class.
The mask branch is trained to predict a binary mask for each object instance, where the mask is aligned with the object's bounding box and has the same size as the region of interest (RoI). The predicted mask is then used to segment the object from the background.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
 **************
 
-To be added soon
+For the dataset handling inside OTX, we use `Dataset Management Framework (Datumaro) <https://github.com/openvinotoolkit/datumaro>`_. For instance segmentation we support `COCO <https://cocodataset.org/#format-data>` and `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_ dataset formats.
```
```suggestion
For the dataset handling inside OTX, we use `Dataset Management Framework (Datumaro) <https://github.com/openvinotoolkit/datumaro>`_. For instance segmentation we support `COCO <https://cocodataset.org/#format-data>`_ and `Pascal-VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_ dataset formats.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
+
+- ``Optimizer``: We use `SGD <https://en.wikipedia.org/wiki/Stochastic_gradient_descent>`_ optimizer with the weight decay set to **1e-4** and momentum set to **0.9**.
+
+- ``Learning rate schedule``: For scheduling training process we use **ReduceLrOnPlataeu** with linear learning rate warmup for **200** iterations. This method monitors a target metric (in our case we use metric on the validation set) and if no improvement is seen for a 'patience' number of epochs, the learning rate is reduced.
```
This `patience` supposed to be highlighted? Because it's rendered in other way
![image](https://user-images.githubusercontent.com/33455576/216109134-d3e095b0-6116-4850-9cf3-3a825d54c6bf.png)

kprokofi replied:
No, it is basically brackets, but I will remove them to not mislead anyone

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
+
+    OTX will recognize train and validation datasets automatically. If you want to emphasize the validation dataset root path, you can simply add the ``--val-data-root <path_to_val_data_root>`` option.
+    If you have your annotations in COCO format, but, for example, have the structure or names differ from the original one - you can directly pass paths to annotation files to the command line with two CLI options: ``--train-ann-files <path_to_train_annotations.json>``, ``--val-ann-files <path_to_val_annotations.json>``.
+    Please, refer to our :doc:`dedicated tutorial <../../../tutorials/base/how_to_train/instance_segmentation>` how to train, validate and optimize instance segmentation model for more details.
```
Add additional enter before this paragraph to make a new point.

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
+.. note::
+
+    OTX will recognize train and validation datasets automatically. If you want to emphasize the validation dataset root path, you can simply add the ``--val-data-root <path_to_val_data_root>`` option.
+    If you have your annotations in COCO format, but, for example, have the structure or names differ from the original one - you can directly pass paths to annotation files to the command line with two CLI options: ``--train-ann-files <path_to_train_annotations.json>``, ``--val-ann-files <path_to_val_annotations.json>``.
```
Also, shouldn't the same note be in the `Semantic Segmentation` page?

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
 **************
 
-To be added soon
+For the dataset handling inside OTX we use `Dataset Management Framework (Datumaro) <https://github.com/openvinotoolkit/datumaro>`_. At this end we support `ADE20K <https://openvinotoolkit.github.io/datumaro/docs/formats/ade20k2020/>`_, `Cityscapes <https://openvinotoolkit.github.io/datumaro/docs/formats/cityscapes/>`_, `Pascal VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_v ...
```
```suggestion
For the dataset handling inside OTX we use `Dataset Management Framework (Datumaro) <https://github.com/openvinotoolkit/datumaro>`_. 

At this end we support `ADE20K <https://openvinotoolkit.github.io/datumaro/docs/formats/ade20k2020/>`_, `Cityscapes <https://openvinotoolkit.github.io/datumaro/docs/formats/cityscapes/>`_, `Pascal VOC <https://openvinotoolkit.github.io/datumaro/docs/formats/pascal_voc/>`_ and `Common Semantic Segmentation <https://openvinotoolkit.github.io/datumaro/docs/formats/common_semantic_segmentation/>`_ data formats.
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
+
+    Please, note, that images should have ``.jpg`` extension and segmentation masks should be ``.png``.
+
+Since this format isn't handle by Datumaro we can start training by passing the training and validation data paths manually:
```
```suggestion
Since this format isn't handled by Datumaro, we can start training by passing the training and validation data paths manually:
```

GalyaZalesskaya commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,145 @@
 ************************
 
-To be added soon
+To solve Semi-supervised learning problem for the semantic segmentation we use the `Mean Teacher algorithm <https://arxiv.org/abs/1703.01780>`_. The basic idea of this approach is to use two models during training: a "student" model, which is the main model being trained, and a "teacher" model, which acts as a guide for the student model.
```
Can you please add a line here: "to learn about Semi-supervised Learning refer to algorithms/index?

GalyaZalesskaya reviewed (commented):
Thank you for your work. Also, you probably added extra copies of the same images in PR in utils/images :)

kprokofi committed:
fix semantic:

kprokofi committed:
fix indents. Reflected comments

kprokofi committed:
fix

kprokofi commented:
@GalyaZalesskaya comments are reflected

JihwanEom commented on docs/source/explanation/algorithms/segmentation/instance_segmentation.rst:
```diff
@@ -1,17 +1,76 @@
++--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+---------------------+-----------------+
+| Template ID                                                                                                                                                                                                                                | Name                       | Complexity (GFLOPs) | Model size (MB) |
++============================================================================================================================================================================================================================================+============================+=====================+=================+
+| `Custom_Counting_Instance_Segmentation_MaskRCNN_EfficientNetB2B <https://github.com/openvinotoolkit/training_extensions/blob/feature/otx/otx/algorithms/detection/configs/instance_segmentation/efficientnetb2b_maskrcnn/template.yaml>`_  | MaskRCNN-EfficientNetB2B   | 68.48               | 13.27           |
```
Could you update links to `develop`?
```suggestion
| `Custom_Counting_Instance_Segmentation_MaskRCNN_EfficientNetB2B <https://github.com/openvinotoolkit/training_extensions/blob/develop/otx/algorithms/detection/configs/instance_segmentation/efficientnetb2b_maskrcnn/template.yaml>`_  | MaskRCNN-EfficientNetB2B   | 68.48               | 13.27           |
```

JihwanEom reviewed (changes requested):
Please replace `feature/otx` with `develop`

sungmanc commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,49 @@
+
+    .. code-block::
+
+        $ otx train --train-data-root <path_to_data_root> --task {CLASSIFICATION, DETECTION, SEGMENTATION, ACTION_CLASSIFICATION, ACTION_DETECTION, ANOMALY_CLASSIFICATION, ANOMALY_DETECTION, ANOMALY_SEGMENTATION, INSTANCE_SEGMENTATION}
```
Current `build` and `train` CLI have different arguments,

For the `otx build` --> `--model`, `--backbone`, `--train_type`, `--task`, ...
For the `otx train` --> `--train/val-data-roots`, `--template`, `--save-model-to`, ...

`otx build --train-data-root {}` --> OK.
`otx build --train-data-root {} --task {}` --> OK.
`otx build --train-data-root {} --template {}` --> Can't

`otx train --train-data-root {}` --> OK.
`otx train --train-data-root {} --task {}` --> Can't
`otx train --train-data-root {} --template {}` --> OK.

kprokofi replied:
I changed `train` to `build` in case when we have occluded datasets

supersoob commented on docs/source/explanation/algorithms/classification/multi_class_classification.rst:
```diff
@@ -4,6 +4,8 @@ Multi-class Classification
 Multi-class Classification is the problem of classifying instances into one of two or more classes. We solve this problem in a common fashion, based on the feature extractor backbone and classifier head that predicts the distribution probability of the categories from the given corpus.
```
It can be minor but I prefer to write `Multi-class classification` rather then `Multi-class Classification` in the body if it's not header. Same as for Hierarchical classification and other parts.

supersoob replied:
One more minor suggestion : I suggest to describe the first sentence to -> "Multi-class classification has a challenge to classify instances into multiple classes."

kprokofi replied:
'one of two' emphasize that we include also binary classification. Then, 'problem' I think is a common word for describing task setting in the literature rather then a 'challenge'. So, I prefer keep it as is. What do you think?

supersoob replied:
Good:)

supersoob commented on docs/source/explanation/algorithms/classification/multi_class_classification.rst:
```diff
@@ -13,19 +15,9 @@ For the supervised training we use the following algorithms components:
 - ``Loss function``: `Influence-Balanced Loss <https://arxiv.org/abs/2110.02444>`_. This is a balancing training method that helps us to solve the imbalanced data problem.
```
"IB loss is a solution for class-imbalance, which avoids overfitting to the majority classes re-weighting the influential samples."

supersoob replied:
In addition, we generally use CrossEntropy just when no new classes are added. Only the case we detect class incremental, IB loss is used as class-imbalance solution.

kprokofi replied:
Wouldn't it be always 'class incremental' since we use snapshot from ImageNet and number of classes will be different from data?
![image](https://user-images.githubusercontent.com/68103869/216402367-2509357f-2208-431d-a05c-c9160ebb92a9.png)

kprokofi replied:
@JihwanEom @supersoob What do you think?

supersoob replied:
Sorry for late comment. Even though we use the imagenet ckpt at the first time training, model classes start off with a clean-slate. So it just use CrossEntropy when there's no difference between model and data classes as stated in if-statement.

kprokofi replied:
I changed it

supersoob commented on docs/source/explanation/algorithms/classification/multi_class_classification.rst:
```diff
@@ -4,6 +4,8 @@ Multi-class Classification
 Multi-class Classification is the problem of classifying instances into one of two or more classes. We solve this problem in a common fashion, based on the feature extractor backbone and classifier head that predicts the distribution probability of the categories from the given corpus.
 For the supervised training we use the following algorithms components:
 
+.. _mcl_cls_supervised_pipeline:
```
Plus, I'm curious if there's a section to explain the sampler we used for incremental solution. If not, I'd like to add here. Especially for multi-class classification it uses specific balanced sampler.

kprokofi replied:
Again, why do we do it silently in the code? I would rather write it in the config
![image](https://user-images.githubusercontent.com/68103869/216465839-07e0a084-4cdd-4700-ae42-2c2d5ea52027.png)

kprokofi replied:
I added description about balanced sampler. Please, check it, because I don't know exactly what it is doing. I took info from doc string

supersoob replied:
TaskAdaptHook needs some dynamically changing args such as new or old classes which can be checked in code. So we re-configure it in stage.py again. But as you said, we can register TaskAdaptHook in config just with base args. Looks like we need some refactoring again.

supersoob commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,155 @@
+
+- ``Learning rate schedule``: For scheduling training process we use **ReduceLROnPlataeu** with linear learning rate warmup for 100 iterations. This method monitors a target metric (in our case we use metric on the validation set) and if no improvement is seen for a 'patience' number of epochs, the learning rate is reduced.
+
+- ``Loss function``: We use standart `Cross Entropy Loss <https://en.wikipedia.org/wiki/Cross_entropy>`_  to train a model.
```
@JihwanEom Do you think we need to explain the ignore management in loss?

kprokofi replied:
Why do we silently substitute losses and other components in config?
![image](https://user-images.githubusercontent.com/68103869/216462093-8edd066e-e1f4-4b45-8813-0fc3d863177a.png)
task_adapt always True for each config-> why can't we directly point which loss we are using in config? Not just `Cross Entropy`, but `CrossEntropyWithIgnore`

kprokofi replied:
Anyway, it is for class incremental -> for incremental learning exists an additional section below which hasn't been filled yet

supersoob replied:
Yeah if we have a additional section for incremental, it would be good to explain in there.

supersoob commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,155 @@
+                ├── 1.jpg
+                ...
+                └── N.jpg
+        └── annotations
```
This should be "masks" for datumaro format. Isn't it? @sungmanc

kprokofi replied:
What is the masks? Proper naming for "images" and "annotations"? Yes, @sungmanc, where can I find exact folders naming for each format?

kprokofi replied:
We can pass `--train-data-root` to train and `--val-data-root` to val and the training will be started

supersoob commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,155 @@
+
+    Please, refer to our :doc:`dedicated tutorial <../../../tutorials/base/how_to_train/semantic_segmentation>` how to train, validate and optimize semantic segmentation model for more details.
+
+
```
Before saying that we support the custom format in seg, I hope we add the description line that we basically support VOC and cityscapes dataset original format in Datumaro.

kprokofi replied:
I'm not fully understand what you mean by that, but I described data formats supported by Datumaro in the 36-41 lines. Even gave a CLI command line how to run training

supersoob commented on docs/source/explanation/algorithms/segmentation/semantic_segmentation.rst:
```diff
@@ -1,23 +1,155 @@
+After training, only the student model is used for prediction.
+
+We utilize the same core algorithm's parameters as for the :ref:`supervised pipeline <semantic_segmentation_supervised_pipeline>`. The main difference is to use of different augmentation pipelines for the labeled and unlabeled data.
+We use only basic augmentations (random flip, random rotate, random crop) for the labeled data and more severe for the unlabeled (the same as we use for the :ref:`supervised pipeline <semantic_segmentation_supervised_pipeline>`).
```
```suggestion
We use only basic augmentations (random flip, random rotate, random crop) for the labeled data and stronger for the unlabeled (color distortion).
```

kprokofi committed:
change everything to develop

kprokofi committed:
remove duplicated imgs

kprokofi committed:
minor

kprokofi committed:
added balance loss description

kprokofi committed:
changed task for build in auto-config

supersoob commented on docs/source/explanation/algorithms/classification/multi_class_classification.rst:
```diff
@@ -1,31 +1,23 @@
 
-- Additionally, we use `No Bias Decay (NBD) <https://arxiv.org/abs/1812.01187>`_ technique and **early stopping** to add adaptability to the training pipeline and prevent overfitting.
-To further enhance the performance of the algorithm in case when we have a small number of data we use `Supervised Contrastive Learning <https://arxiv.org/abs/2004.11362>`_. More specifically, we train a model with two heads: classification head with Influence-Balanced Loss and SupCon head with `Barlow Twins loss <https://arxiv.org/abs/2103.03230>`_.
+- Additionally, we use `No Bias Decay (NBD) <https://arxiv.org/abs/1812.01187>`_ technique and **early stopping** to add adaptability to the training pipeline and prevent overfitting. Besides this we use `Balanced Sampler <https://github.dev/openvinotoolkit/training_extensions/blob/develop/otx/mpa/modules/datasets/samplers/balanced_sampler.py#L11>`_ to create an efficient batch that allow reducin ...
```
``
```suggestion
- Additionally, we use `No Bias Decay (NBD) <https://arxiv.org/abs/1812.01187>`_ technique and **early stopping** to add adaptability to the training pipeline and prevent overfitting. Besides this we use `Balanced Sampler <https://github.dev/openvinotoolkit/training_extensions/blob/develop/otx/mpa/modules/datasets/samplers/balanced_sampler.py#L11>`_ to create an efficient batch that consists of balanced samples over classes, reducing the iteration size as well.
```

kprokofi replied:
Thank you!

kprokofi committed:
fix balanced sampler

kprokofi committed:
refactored CLI interface

kprokofi committed:
minor

kprokofi commented:
@JihwanEom , @supersoob , @sungmanc I reflected comments (except one with IB loss, need additional discussion)

sungmanc commented on docs/source/explanation/additional_features/auto_configuration.rst:
```diff
@@ -1,4 +1,57 @@
+        $ otx build --train-data-root <path_to_data_root>
+
+Moreover, our dataset can have no train/val splits at all. The Datumaro manager integrated into OTX will handle it on its own.
+It will recognize the task by analyzing the dataset and if there is no splits for the validation - Datumaro will do a random auto-split, memorizing this split to further use it with ``otx eval`` and ``otx optimize``.
```
One small suggestion, 

Since `otx build` will makes the `train` and `val` datasets, users can't directly use these datasets to `otx eval` which is required `test` dataset. In the case of `otx optimize`,there is no problem. 

So, how about converting like below?

It will recognize the task by analyzing the dataset and if there is no splits for the validation - Datumaro will do a random auto-split, saving this split to the workspace. It could be used with ``otx optimize`` or ``otx train``.

kprokofi replied:
Okay, maybe this is better, I changed it.

kprokofi committed:
fix loss description for multi-class classification

kprokofi committed:
fix auto-config

kprokofi commented:
@JihwanEom , @supersoob , @GalyaZalesskaya , @sungmanc 
Please, approve if there is no other comments. I resolved all current comments.

sungmanc reviewed (approved):
This is a quite nice work! Thanks a lot !

supersoob reviewed (approved):
It looks very nice. Thank you:)

goodsong81 reviewed (approved):
What a documentation Maestro!
